rsyslog/go/src/github.com/mmexternal/undefined_field/undefined_field.go

```diff
@@ -0,0 +1,239 @@
+				undefined_to_string = rval
+			}
+		case "CDM_UNDEFINED_DOT_REPLACE_CHAR":
+			undefined_dot_replace_char = keyVal[1]
+		case "CDM_UNDEFINED_MAX_NUM_FIELDS":
+			var rval64 int64
+			if rval64, err = strconv.ParseInt(keyVal[1], 10, 64); err == nil {
+				undefined_max_num_fields = rval64
+			}
+		}
+	}
+
+	if !use_undefined && tmp_keep_empty_fields == "" && !undefined_to_string && undefined_dot_replace_char == "UNUSED" {
+		noaction = true
+	}
+	if undefined_max_num_fields == -1 {
+		undefined_max_num_fields = 9223372036854775807
+	}
+	if use_undefined {
+		tmp_basic := strings.Split(basic_keep_fields, ",")
+		tmp_default := strings.Split(default_keep_fields, ",")
+		tmp_extra := strings.Split(extra_keep_fields, ",")
+		keep_fields = append(tmp_basic, tmp_default...)
+		keep_fields = append(keep_fields, tmp_extra...)
+	}
+	keep_empty_fields = strings.Split(tmp_keep_empty_fields, ",")
+
+	// var replacer *strings.Replacer = nil
+	if undefined_dot_replace_char != "UNUSED" {
+		replacer = strings.NewReplacer(".", undefined_dot_replace_char)
+	}
+
+	fmt.Fprintln(logfile, "mmexternal: use_undefined: ", use_undefined)
+	fmt.Fprintln(logfile, "mmexternal: default_keep_fields: ", default_keep_fields)
+	fmt.Fprintln(logfile, "mmexternal: extra_keep_fields: ", extra_keep_fields)
+	fmt.Fprintln(logfile, "mmexternal: undefined_name: ", undefined_name)
+	fmt.Fprintln(logfile, "mmexternal: keep_empty_fields: ", keep_empty_fields)
+	fmt.Fprintln(logfile, "mmexternal: undefined_to_string: ", undefined_to_string)
+	fmt.Fprintln(logfile, "mmexternal: undefined_dot_replace_char: ", undefined_dot_replace_char)
+	fmt.Fprintln(logfile, "mmexternal: undefined_max_num_fields: ", undefined_max_num_fields)
+	fmt.Fprintln(logfile, "mmexternal: noaction: ", noaction)
+}
+
+func contains(slice []string, item string) bool {
+	for _, n := range slice {
+		if item == n {
+			return true
+		}
+	}
+	return false
+}
+
+func replaceDotAndSkipEmpty(copyTop *map[string]interface{}, jsonMap map[string]interface{}, domove bool) map[string]interface{} {
+	cp := make(map[string]interface{})
+	if *copyTop == nil {
+		*copyTop = cp
+	}
+	for key, value := range jsonMap {
+		if undefined_dot_replace_char != "UNUSED" {
+			// replace '.' with specified char (e.g., '_')
+			key = replacer.Replace(key)
+		}
+		// skip empty or not?
+		valuemap, ok := value.(map[string]interface{})
+		if !contains(keep_empty_fields, key) {
+			if !ok && (value == nil || len(value.(string)) == 0) || ok && len(valuemap) == 0 {
+				continue
+			}
+		}
+		// use_undefined and key is not in keep_fields and less than unmdefined_max_num_fields?
+		if domove && use_undefined && !contains(keep_fields, key) && undefined_max_num_fields > 0 {
+			tmpcp := *copyTop
+			if tmpcp[undefined_name] == nil {
+				subcp := make(map[string]interface{})
+				tmpcp[undefined_name] = subcp
+			}
+			if ok {
+				rval := replaceDotAndSkipEmpty(copyTop, valuemap, false)
```

Comment:
why do you need to recurse here?  we are only concerned with the top level fields - are you trying combine both the undefined field handling, and checking to see if the nested structure is empty?  If so, I think it would be better to have a separate `isempty` or `delempty` function like https://github.com/ViaQ/fluent-plugin-viaq_data_model/blob/master/lib/fluent/plugin/filter_viaq_data_model.rb#L232

Reply:
> I think you need to change the undefined_dot_replace_char check to only check the top level fields

you also want to make sure you use the original `key` for testing `keep_empty_fields` and `keep_fields` - the new `key` should only be used for `cp[newkey] = rval`, and only for top level undefined fields - that is, if `key` is a defined field (in `keep_fields`), or `key` is a recursively nested field, we should not do the replacement on it.  Maybe the code already does that, but I can't tell.